dvc/remote/ssh/pool.py

```diff
@@ -0,0 +1,50 @@
+from collections import deque
+from contextlib import contextmanager
+from funcy import memoize
+
+from .connection import SSHConnection
+
+
+@contextmanager
+def ssh_connection(*conn_args, **conn_kwargs):
+    pool = get_ssh_pool(*conn_args, **conn_kwargs)
+    conn = pool.get_connection()
+    try:
+        yield conn
+    except BaseException:
+        conn.close()
+    else:
+        pool.release(conn)
+
+
+@memoize
+def get_ssh_pool(*conn_args, **conn_kwargs):
+    return SSHPool(conn_args, conn_kwargs)
+
+
+class SSHPool(object):
+    def __init__(self, conn_args, conn_kwargs):
+        self._conn_args = conn_args
+        self._conn_kwargs = conn_kwargs
+        self._conns = deque()
+        self._closed = False
+
+    def __del__(self):
+        self.close()
+
+    def close(self):
+        while self._conns:
+            self._conns.pop().close()
+        self._closed = True
+
+    def get_connection(self):
+        try:
+            return self._conns.popleft()
```

Comment:
So we will have the same SSHPool even if we have multiple connections to different remotes? The password thingy at least has a scope of 1 remote instance, but this pool is global, so if you run one instance (to remote1) and then second instance to remote2, the remote2 will try to use the same connection to remote1, right?

Reply:
Pools are keyed by all the args and kwargs passed to `SSHConnection`, so these would be different pools.